iml-util/util.rs

```diff
@@ -56,6 +56,76 @@ impl<A, B, C, D, E> Pick<A, B> for (A, B, C, D, E) {
     }
 }
 
+pub mod tokio_utils {
+    use futures::{future::Either, Stream, TryStreamExt};
+    use std::{
+        convert::TryFrom,
+        env, io,
+        os::unix::{io::FromRawFd, net::UnixListener as NetUnixListener},
+        pin::Pin,
+    };
+    use tokio::{
+        io::{AsyncRead, AsyncWrite},
+        net::{TcpListener, UnixListener},
+    };
+
+    pub trait Socket: AsyncRead + AsyncWrite + Send + 'static + Unpin {}
+    impl<T: AsyncRead + AsyncWrite + Send + 'static + Unpin> Socket for T {}
+
+    /// Given an environment variable that resolves to a port,
+    /// Return a stream containing `TcpStream`s that have been erased to
+    /// `AsyncRead` + `AsyncWrite` traits. This is useful when you won't know which stream
+    /// to choose at runtime
+    pub async fn get_tcp_stream(
+        port: String,
+    ) -> Result<impl Stream<Item = Result<Pin<Box<dyn Socket>>, io::Error>>, io::Error> {
+        let addr = format!("127.0.0.1:{}", port);
+
+        tracing::debug!("Listening over tcp port {}", port);
+
+        let s = TcpListener::bind(&addr)
+            .await?
+            .incoming()
+            .map_ok(|x| -> Pin<Box<dyn Socket>> { Box::pin(x) });
+
+        Ok(s)
+    }
+
+    /// Returns a stream containing `UnixStream`s that have been erased to
+    /// `AsyncRead` + `AsyncWrite` traits. This is useful when you won't know which stream
+    /// to choose at runtime
+    pub fn get_unix_stream(
+    ) -> Result<impl Stream<Item = Result<Pin<Box<dyn Socket>>, io::Error>>, io::Error> {
+        let addr = unsafe { NetUnixListener::from_raw_fd(3) };
+
+        tracing::debug!("Listening over unix domain socket");
```

Comment:
Just a question: what is opening this unix socket? It must be a file somewhere?

Reply:
This is opened when we use as a systemd service with socket-activation. ex: 

https://github.com/whamcloud/integrated-manager-for-lustre/blob/master/iml-action-runner.socket

and

https://github.com/whamcloud/integrated-manager-for-lustre/blob/master/iml-action-runner.service